[gwctl ] feat: `gwctl get backends` support

yashvardhan-kukreja opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:

1. If this is your first time contributing to Gateway API, please read our
   developer guide (https://gateway-api.sigs.k8s.io/contributing/devguide/)
   and our community page (https://gateway-api.sigs.k8s.io/contributing/).
2. If this is your first time contributing to a Kubernetes project, please read
   our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
3. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
4. If you want *faster* PR reviews, read how:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
6. If this PR includes a new GEP please make sure you've followed the process
   outlined in our GEP overview, as this will help the community to ensure the
   best chance of positive outcomes for your proposal:
   https://gateway-api.sigs.k8s.io/geps/overview/#process
-->

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind gep
/kind test

Optionally add one or more of the following kinds if applicable:
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/area conformance
-->

**What this PR does / why we need it**:

/kind feature

This PR introduces the `gwctl get backends` subcommand in favor of the requirements of this [GEP](https://deploy-preview-2723--kubernetes-sigs-gateway-api.netlify.app/geps/gep-2722/)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #2866 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
Yes, it introduces the `gwctl get backends` command to support returning backends to the console.
For example:
```release-note
gwctl get backends

NAMESPACE  NAME       TYPE     REFERRED BY ROUTES                                   AGE   POLICIES
default    foo-svc-0  Service  default/foo-httproute-1                              3d    2
ns1        foo-svc-1  Service  default/qmn-httproute-100,default/foo-httproute-1    2d    3
ns2        foo-svc-2  Service  default/qmn-httproute-100,ns1/bar-route-21 + 1 more  36h   3
ns3        foo-svc-3  Service  default/qmn-httproute-100,ns1/bar-route-21           24h   2
ns3        foo-svc-4  Service  None                                                 5d8h  0
```

yashvardhan-kukreja commented:
HI @mlavacca @gauravkghildiyal , can one of you attach the ok-to-test label to this PR?

Thanks!

Devaansh-Kumar commented on gwctl/pkg/printer/backends.go:
```diff
@@ -28,7 +34,96 @@ import (
+		associatedDistinctPolicies := map[string]any{}
+
+		// the nested loops could've been made more efficient (on textbook) by using map between httpRouteNodes and backendRefs
+		// and leveraging the O(1) lookups but that seemed to sacrifice the code readability/structure, so didn't choose to do so
```
According to me, creating maps would be helpful even if we lose some readability.

yashvardhan-kukreja replied:
I personally don't think so for the following reasons:
- the difference in runtime would be negligible majorly,  almost unrecognisable the end-user's experience.
- For significantly large use-cases (which would be rare), the map would lead to a much larger memory footprint.

Thank you for your suggestion but I'd rather incline towards the better readability and maintainability of this subroutine than giving in to a potentially premature optimization.

Rest let's wait for the maintainers to hear their opinions.

Cheers!

gauravkghildiyal commented:
/assign
/ok-to-test

gauravkghildiyal commented on gwctl/pkg/printer/backends.go:
```diff
@@ -28,7 +34,96 @@ import (
+					remainderHttpRoutes += 1
+				}
+
+				for _, effectivePolicy := range httpRouteNode.EffectivePolicies {
```
The space around inherited policies is somewhat still under changes. For the time being, let's simply imply the count of policies to be the policies that are directly attached to this Backend. This would mean simply counting the number of elements in the `BackendNode.Policies`

gauravkghildiyal commented on gwctl/pkg/printer/backends.go:
```diff
@@ -28,7 +34,96 @@ import (
+
+		// the nested loops could've been made more efficient (on textbook) by using map between httpRouteNodes and backendRefs
+		// and leveraging the O(1) lookups but that seemed to sacrifice the code readability/structure, so didn't choose to do so
+		for _, httpRouteNode := range httpRouteNodes {
```
Did we perhaps overlook this or is there some reason to not use the mappings already provided through the resourceModel (i.e. using the `BackendNode.HTTPRoutes`? If there's a bug in those mappings, we should try and fix this (separately, if required)

https://github.com/kubernetes-sigs/gateway-api/blob/29e68bffffb9af568e35545305d78d0001a1a0f7/gwctl/pkg/resourcediscovery/nodes.go#L215-L228

gauravkghildiyal reviewed (commented):
Thanks @yashvardhan-kukreja -- this one is almost ready. Just a few clarifications.

yashvardhan-kukreja commented:
Hi @gauravkghildiyal , apologies for the late catchup on this PR, I was going through my exams.
Now that I am done with my exams, just wanted to give you a heads up that I'll catchup with your comments.

Thanks :)

yashvardhan-kukreja commented:
Hi @gauravkghildiyal , I have addressed your comments and broken the PR into 3 commits, separating the main feature, the associated tests, and some linty nits.

Thank you

gauravkghildiyal commented on gwctl/pkg/printer/backends_test.go:
```diff
@@ -0,0 +1,485 @@
+	}
+	resourceModel, err := discoverer.DiscoverResourcesForBackend(resourcediscovery.Filter{})
+	if err != nil {
+		t.Fatalf("Failed to construct resourceModel: %v", resourceModel)
```
```suggestion
		t.Fatalf("Failed to construct resourceModel: %v", err)
```

(I know we have the same issue in other places, and we'll get to it independently, or when we  make changes to those tests)

gauravkghildiyal commented on gwctl/pkg/printer/backends.go:
```diff
@@ -28,7 +36,101 @@ import (
+				remainderHTTPRoutes++
+			}
+
+			// TODO(yashvardhan-kukreja): change this to httpRouteNode.EffectivePolicies once the work around inherited policies is matured
```
You should be counting the policies which are directly attached to the Backend: `BackendNode.Policies` (might need to adjust the tests as well)

yashvardhan-kukreja replied:
@gauravkghildiyal I tried doing this but the output seemed a bit weird.

None of the backendRefs got any policies associated with them with the above method.

Can you give an example of a directly attached policy to a backend? i.e. which resource can I apply to simulate directly attaching policy to a backend.

gauravkghildiyal reviewed (commented):
Thanks. This is almost ready!

yashvardhan-kukreja commented:
Addressed your comments @gauravkghildiyal

gauravkghildiyal commented on gwctl/pkg/printer/gateways.go:
```diff
@@ -24,12 +24,13 @@ import (
 	"strings"
 	"text/tabwriter"
 
+	"sigs.k8s.io/yaml"
```
nit: Are these diffs (including httproutes.go, policies.go) a result of some linter complaining? If not, can we minimize such changes.

gauravkghildiyal reviewed (commented):
Just one nit. 

/approve

gauravkghildiyal commented:
/lgtm
